src/providers/postgres/qgspostgresfeatureiterator.cpp

```diff
@@ -506,10 +506,60 @@ QString QgsPostgresFeatureIterator::whereClauseRect()
   bool castToGeometry = mSource->mSpatialColType == SctGeography ||
                         mSource->mSpatialColType == SctPcPatch;
 
-  QString whereClause = QStringLiteral( "%1%2 && %3" )
-                        .arg( QgsPostgresConn::quotedIdentifier( mSource->mBoundingBoxColumn ),
-                              castToGeometry ? "::geometry" : "",
-                              qBox );
+  QString whereClause;
+  if ( mSource->mSpatialColType == SctTopoGeometry &&
+       mSource->mTopoLayerInfo.layerLevel == 0 &&
+       mSource->mTopoLayerInfo.hasIndexOnTopoGeometryID )
+  {
+    whereClause = QStringLiteral( R"SQL(
+      id(%1) in (
+        WITH qb AS (
+          select %2 b
+        ),
+        elems AS (
+
+          -- TODO: skip if layer cannot be composed by nodes
+          SELECT ARRAY[n.node_id,1] te
```

Comment:
Nit: you might end up rewriting this anyway, but returning two columns there is probably easier for readers, and for Postgres too.

Reply:
Thanks, ARRAY construction removed. For the record, readers might have recognized I was building (uselessly) "TopoElement" object -- https://postgis.net/docs/topoelement.html